### ISSUE TITLE
Uncalled function checks do not error on incomplete curried function expressions

### ISSUE THREAD

**Issue by rene-leanix** [microsoft/TypeScript#42600]
# Bug Report

### 🔎 Search Terms

- "Uncalled function checks"

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about "Uncalled function checks"

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAlhACiCcpwG4FMYF4YAUAHgFwxgCuAtgEZYBOAlHgHyECeZltDzubRGAGoY7GGwAMAbgBQMuADNCCZKnTZmAbwC+cxcqQo0mLAQC0ARkZbtQA)

### 💻 Code

```ts
const isPositive = (x: number) => (y: number) => x + y > 0;

if (isPositive) {}
    ^^^ error shown

if (isPositive(-1)) {}
    ^^^ error missing
```

### 🙁 Actual behavior

Uncalled function conditions are not shown as errors, if the function is returned from a function call, e.g. for a curried function with incomplete arguments. (Curried functions are particular tricky when used in conditions, as it may not be immediately transparent how many arguments they require.)

### 🙂 Expected behavior

Both uncalled function checks are marked as errors.

**Comment by RyanCavanaugh** [comment]
The uncalled function error is *extremely* conservative and only appears when we have very high confidence it's not a false positive. Any time we tried to enforce this more strictly there were too many false positives to ship it. For this particular case we don't have any good way to detect when it's intentional and thus need to let it be. I believe there's an lint rule available for always forcing boolean coercion you can use instead.
